clang/test/CIR/CodeGen/builtins-memory.c

```diff
@@ -0,0 +1,56 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
+// RUN: FileCheck %s --check-prefix=CIR --input-file=%t.cir
+
+typedef __SIZE_TYPE__ size_t;
+void test_memcpy_chk(void *dest, const void *src, size_t n) {
+  // CIR-LABEL: cir.func @test_memcpy_chk
+  // CIR:         %[[#DEST:]] = cir.alloca {{.*}} ["dest", init]
+  // CIR:         %[[#SRC:]] = cir.alloca {{.*}} ["src", init]
+  // CIR:         %[[#N:]] = cir.alloca {{.*}} ["n", init]
+
+  // memcpy should be emitted when the count is known to fit the buffer size.
+
+  // CIR: %[[#DEST_LOAD:]] = cir.load %[[#DEST]]
+  // CIR: %[[#SRC_LOAD:]] = cir.load %[[#SRC]]
+  // CIR: %[[#COUNT:]] = cir.const #cir.int<8>
+  // CIR: cir.libc.memcpy %[[#COUNT]] bytes from %[[#SRC_LOAD]] to %[[#DEST_LOAD]]
+  __builtin___memcpy_chk(dest, src, 8, 10);
+
+  // CIR: %[[#DEST_LOAD:]] = cir.load %[[#DEST]]
+  // CIR: %[[#SRC_LOAD:]] = cir.load %[[#SRC]]
+  // CIR: %[[#COUNT:]] = cir.const #cir.int<10>
+  // CIR: cir.libc.memcpy %[[#COUNT]] bytes from %[[#SRC_LOAD]] to %[[#DEST_LOAD]]
+  __builtin___memcpy_chk(dest, src, 10, 10);
+
+  // A count known to be smaller than the size should call the libc function.
+
+  // CIR: %[[#DEST_LOAD:]] = cir.load %[[#DEST]]
+  // CIR: %[[#SRC_LOAD:]] = cir.load %[[#SRC]]
+  // CIR: %[[#COUNT:]] = cir.const #cir.int<10>
+  // CIR: %[[#SIZE:]] = cir.const #cir.int<8>
+  // CIR: cir.call @__memcpy_chk(%[[#DEST_LOAD]], %[[#SRC_LOAD]], %[[#COUNT]], %[[#SIZE]])
+  __builtin___memcpy_chk(dest, src, 10lu, 8lu);
+
+  // Either the count or size not being a constant should call the libc function.
+
+  // CIR: %[[#DEST_LOAD:]] = cir.load %[[#DEST]]
+  // CIR: %[[#SRC_LOAD:]] = cir.load %[[#SRC]]
+  // CIR: %[[#N_LOAD:]] = cir.load %[[#N]]
+  // CIR: %[[#SIZE:]] = cir.const #cir.int<10>
+  // CIR: cir.call @__memcpy_chk(%[[#DEST_LOAD]], %[[#SRC_LOAD]], %[[#N_LOAD]], %[[#SIZE]])
```

Comment:
Are all the `__memcpy_chk` and `libc` check calls reversed? Isn't this supposed to be `cir.libc.memcpy` per the comment on L34 or is the comment reversed?

Reply:
Ah, "libc function" is ambiguous here. I meant the actual underlying function from the library, not the Clang intrinsic. I'll reword it.

Reply:
Nice catch Vincent. Shoaib, see comment above.